### PR TITLE
UAR-1432 Support for adding a ceased date to trusts

### PR DESCRIPTION
**Pull request by mwestacott**
### JIRA link

https://companieshouse.atlassian.net/browse/UAR-1432

### Change description

* New DTO and DAO fields with getters and setters defined
* Trust validation logic refactored and extended to check the new field
* Unit tests added and updated

### Work checklist

- [X] Tests added where applicable

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

**Comment by mwestacott** [comment]
I need to address the code smell...

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dao/trust/TrustDataDao.java [hunk 1]
```diff
@@ -16,6 +16,9 @@ public class TrustDataDao {
     @Field("creation_date")
     private LocalDate creationDate;
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dao/trust/TrustDataDao.java:18]: Needs to be changed to `ceased_date` to be consistent with the web field name. Here and elsewhere.

Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/dao/trust/TrustDataDao.java:18]: Done.

**Comment by mwestacott** [comment]
> I need to address the code smell...

Done.

**Comment by mwestacott** [comment]
Making this a draft review again as there are story changes which will effect the API validation.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/OverseasEntitySubmissionDtoValidator.java [hunk 2]
```diff
@@ -121,7 +121,7 @@ private void validateFullCommonDetails(OverseasEntitySubmissionDto overseasEntit
 
     private void validateTrustDetails(OverseasEntitySubmissionDto overseasEntitySubmissionDto, Errors errors, String loggingContext) {
         if (isTrustWebEnabled && !CollectionUtils.isEmpty(overseasEntitySubmissionDto.getTrusts())) {
-            trustDetailsValidator.validate(overseasEntitySubmissionDto.getTrusts(), errors, loggingContext);
+            trustDetailsValidator.validate(overseasEntitySubmissionDto, errors, loggingContext);
             if (!overseasEntitySubmissionDto.isForUpdateOrRemove()) {
                 trustIndividualValidator.validate(overseasEntitySubmissionDto.getTrusts(), errors, loggingContext);
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/OverseasEntitySubmissionDtoValidator.java:126]: (Probably) not an issue, but I don't understand why lines 126 to 128 are only run for a registration. Decision taken by Kainos when API validation for Update was introduced. Would be happy to hear any theories on this...

Comment (markpit) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/OverseasEntitySubmissionDtoValidator.java:126]: Maybe Andre might remember?

Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/OverseasEntitySubmissionDtoValidator.java:126]: @AndreKramer7654 does this ring any bells?

In the meantime, I'm going to add a 'be aware' comment to the code, just to make this a bit clearer.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java [hunk 3]
```diff
@@ -66,17 +71,69 @@ private boolean validateName(String trustName, Errors errors, String loggingCont
...
+                DateValidators.isCeasedDateOnOrAfterCreationDate(trustCeaseDate, creationDate, qualifiedFieldName, errors, loggingContext);
+            }
+        }
+    }
+
+    private boolean isTrustStillRequired(OverseasEntitySubmissionDto overseasEntitySubmissionDto, TrustDataDto trustDataDto) {
+        // Use the trust id on TrustDataDto to whizz through all the Individual and Corporate BOs that are associated
+        // with this trust. As soon as a matching BO is found where cease date is null and a trust NOC is set then this
+        // indicates that the trust is still required
+
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java:112]: you could pull out the trustId null check from both ifs and return early
`if (trustDataDto.getTrustId() == null) {
        return false; // No trust ID, so trust is not required
    }`

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java [hunk 4]
```diff
@@ -66,17 +71,69 @@ private boolean validateName(String trustName, Errors errors, String loggingCont
...
+                ApiLogger.infoContext(loggingContext, errorMessage);
+            }
+        } else {
+            final LocalDate creationDate = trustDataDto.getCreationDate();
+
+            // Cease date of the trust cannot be 'null' as there are no longer any Individual or Corporate BOs associated with this trust
+
+            if (UtilsValidators.isNotNull(trustCeaseDate, qualifiedFieldName, errors, loggingContext)) {
+                DateValidators.isDateInPast(trustCeaseDate, qualifiedFieldName, errors, loggingContext);
+                DateValidators.isCeasedDateOnOrAfterCreationDate(trustCeaseDate, creationDate, qualifiedFieldName, errors, loggingContext);
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java:103]: There's no AC for checking if ceased date is on or after creation date, the web doesn't check this either

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java [hunk 5]
```diff
@@ -21,13 +23,16 @@
 @Component
 public class TrustDetailsValidator {
 
-    public Errors validate(List<TrustDataDto> trustDataDtoList, Errors errors, String loggingContext) {
+    public Errors validate(OverseasEntitySubmissionDto overseasEntitySubmissionDto, Errors errors, String loggingContext) {
+
+        List<TrustDataDto> trustDataDtoList = overseasEntitySubmissionDto.getTrusts();
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java:28]: Not sure if this is getting the trusts from the correct location, as they will initially be under update-review trusts before being copied to submission/trusts

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java (hunk 4 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java:103]: I think AC4 and this line of [this](https://companieshouse.atlassian.net/browse/UAR-1425?focusedCommentId=236545) comment covers it:

**_past date that is before the date the trust was created = invalid_**

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java (hunk 3 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java:112]: 👍 will implement.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java (hunk 5 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java:28]: I think it's fine. I've essentially not changed where these are being picked up from, I'm simply now getting them from the same DTO object but a little bit later on. I can explain...

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java (hunk 3 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java:112]: Done.

**Review by markpit — APPROVED** [review]

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java [hunk 6]
```diff
@@ -66,17 +71,71 @@ private boolean validateName(String trustName, Errors errors, String loggingCont
...
+    private void validateCeaseDate(OverseasEntitySubmissionDto overseasEntitySubmissionDto,
+                                   TrustDataDto trustDataDto,
+                                   Errors errors,
+                                   String loggingContext) {
+        final String qualifiedFieldName = getQualifiedFieldName(OverseasEntitySubmissionDto.TRUST_DATA, TrustDataDto.CEASED_DATE_FIELD);
+        final LocalDate trustCeaseDate = trustDataDto.getCeasedDate();
+
+        if (isTrustStillRequired(overseasEntitySubmissionDto, trustDataDto)) {
+            // Cease date of the trust must be 'null' as there are still Individual and/or Corporate BOs associated with this trust
+
```
Comment (dlloyd1-ch) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java:90]: Worth removing the blank line between the comment and the code it refers to.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java [hunk 7]
```diff
@@ -66,17 +71,71 @@ private boolean validateName(String trustName, Errors errors, String loggingCont
...
+            return true;
+        }
+
+        return (overseasEntitySubmissionDto.getBeneficialOwnersCorporate() != null
+                && overseasEntitySubmissionDto.getBeneficialOwnersCorporate().stream().anyMatch(
+                        boCorporateDto -> boCorporateDto.getTrustIds() != null
+                            && boCorporateDto.getTrustIds().contains(trustDataDto.getTrustId())
+                            && boCorporateDto.getCeasedDate() == null
+                            && boCorporateDto.getTrusteesNatureOfControlTypes() != null
+                            && !boCorporateDto.getTrusteesNatureOfControlTypes().isEmpty()));
```
Comment (dlloyd1-ch) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java:133]: Shouldn't this just return false? Would it get past the previous block if it returned true? If so, do you need the previous block?

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java [hunk 8]
```diff
@@ -66,17 +71,71 @@ private boolean validateName(String trustName, Errors errors, String loggingCont
...
+            // Cease date of the trust cannot be 'null' as there are no longer any Individual or Corporate BOs associated with this trust
+
+            if (UtilsValidators.isNotNull(trustCeaseDate, qualifiedFieldName, errors, loggingContext)) {
+                DateValidators.isDateInPast(trustCeaseDate, qualifiedFieldName, errors, loggingContext);
+                DateValidators.isCeasedDateOnOrAfterCreationDate(trustCeaseDate, creationDate, qualifiedFieldName, errors, loggingContext);
+            }
+        }
+    }
+
+    private boolean isTrustStillRequired(OverseasEntitySubmissionDto overseasEntitySubmissionDto, TrustDataDto trustDataDto) {
```
Comment (dlloyd1-ch) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java:108]: Do you need to pass in the trustDto given only the trustId is being used?

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java (hunk 6 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java:90]: 👍 will do so on line 98 too.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java (hunk 8 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java:108]: Good point. No, I'll restrict the width of the interface.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java (hunk 7 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/TrustDetailsValidator.java:133]: Block needed as discussed but I'll make things clearer with the comments.

**Review by dlloyd1-ch — APPROVED** [review]

**Review by markpit — APPROVED** [review]

**Review by markpit — APPROVED** [review]

**Pull request merged**